IMOD/manpages/mtsmooth.man
.na
.nh
.TH mtsmooth 1 4.6.34 IMOD
.SH NAME
mtsmooth - to smooth a model of MTs
.SH SYNOPSIS
mtsmooth
.SH DESCRIPTION
Mtsmooth will take a model of tracked microtubules and smooth the
trajectory of each MT independently.  It will make a much nicer
model for presentation than any kind of transformations will.
However, it will not necessarily preserve the spacing between MTs on
any given section.  Such a smoothed model should thus NOT be used for
any kind of analysis that depends on the separations between MTs.
.P
The program operates by considering each point of each MT in turn.
It fits a polynomial to the nearest N points to that point in the MT,
where N is a number specified by the user.  (More precisely, it fits
a polynomial to X coordinates as a function of Z, and a separate
polynomial to the Y coordinates as a function of Z.)  It replaces the
given point with the coordinates of the fitted polynomial at that
position in Z.  The order of the polynomials (1 for a straight line,
2 for a parabola, etc.,) is also chosen by the user.
.P
Entries to the program:
.P
Name of input model file
.P
Name of output model file
.P
Number of points to fit polynomials to, and the order of the
polynomials.  Fitting to 5 or 7 points, with an order of 2, is
recommended.
.P
Minimum object length for fitting polynomials of the chosen order;
objects shorter than this will be fit with a simple straight line.
A value of 5 is recommended for this parameter.
.SH HISTORY
.nf
Written by David Mastronarde, 1/13/93
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/mtsmooth.f
*       * * * * MTSMOOTH * * * * *
c       
c       MTSMOOTH will take a model of tracked microtubules and smooth the
c       trajectory of each MT independently.  It will make a much nicer
c       model for presentation than any kind of transformations will.
c       However, it will not necessarily preserve the spacing between MTs on
c       any given section.  Such a smoothed model should thus NOT be used for
c       any kind of analysis that depends on the separations between MTs.
c       
c       The program operates by considering each point of each MT in turn.
c       It fits a polynomial to the nearest N points to that point in the MT,
c       where N is a number specified by the user.  (More precisely, it fits
c       a polynomial to X coordinates as a function of Z, and a separate
c       polynomial to the Y coordinates as a function of Z.)  It replaces the
c       given point with the coordinates of the fitted polynomial at that
c       position in Z.  The order of the polynomials (1 for a straight line,
c       2 for a parabola, etc.,) is also chosen by the user.
c       
c       Entries to the program:
c       
c       Name of input model file 
c       
c       Name of output model file
c       
c       Number of points to fit polynomials to, and the order of the
c       polynomials.  Fitting to 5 or 7 points, with an order of 2, is
c       recommended.
c       
c       Minimum object length for fitting polynomials of the chosen order;
c       objects shorter than this will be fit with a simple straight line.
c       A value of 5 is recommended for this parameter.
c       
c       David Mastronarde  1/13/93
c       
      use fortmodel
      parameter (limsec=1000)
      real*4 xmt(limsec),ymt(limsec),zmt(limsec),zfit(1000),slop(50)
c       
      character*80 modelfile,newmodel
      logical exist,readw_or_imod
c       
      call setExitPrefix('ERROR: MTSMOOTH - ')
91    write(*,'(1x,a,$)')'Name of input model file: '
      read(*,'(a)')modelfile
c       
c       read in the model
c       
75    exist=readw_or_imod(modelfile)
      if(.not.exist)go to 91
c       
      write(*,'(1x,a,$)')'Name of output model file: '
      read(*,'(a)')newmodel
c       
      write(*,'(1x,a,$)')
     &    'Number of points to fit to, order of polynomial to fit: '
      read(5,*)nfitpt,iorder
      write(*,'(1x,a,$)')'Minimum number of points required to '//
     &    'fit polynomial rather than line: '
      read(5,*)minfit
c       
      do iobj=1,max_mod_obj
        ninobj=npt_in_obj(iobj)
        if(ninobj.gt.2)then
          ibase=ibase_obj(iobj)
          nzchange=0
          do ind=1,ninobj
            ipt=abs(object(ibase+ind))
            xmt(ind)=p_coord(1,ipt)
            ymt(ind)=p_coord(2,ipt)
            zmt(ind)=p_coord(3,ipt)
            if(ind.gt.1)then
              if(zmt(ind).ne.zmt(ind-1))nzchange=nzchange+1
            endif
          enddo
          if(nzchange.ge.ninobj-2)then
            do ind=1,ninobj
              ipt=abs(object(ibase+ind))
              jfhi=min(ind+nfitpt/2,ninobj)
              jflo=max(1,jfhi+1-nfitpt)
              jfhi=min(ninobj,jflo+nfitpt-1)
              nfit=jfhi+1-jflo
              iord=iorder
              if(nfit.lt.minfit)iord=1
              iz=1
              do jf=jflo,jfhi
                zfit(iz)=zmt(jf)-zmt(ind)
                iz=iz+1
              enddo
              call polyfit(zfit,xmt(jflo),nfit,iord,slop,bint)
              p_coord(1,ipt)=bint
              call polyfit(zfit,ymt(jflo),nfit,iord,slop,bint)
              p_coord(2,ipt)=bint
            enddo
          endif
        endif
      enddo
c       
      call write_wmod(newmodel)
      call exit(0)
      end
